### ISSUE TITLE
Cb ThreatHunter _search() function does not yield the correct num_available results

### ISSUE THREAD

**Issue by crahan** [carbonblack/cbapi-python#239]
The current implementation of the `_search()` function in the Query class in `cbapi/psc/threathunter/query.py` yields an incorrect number of results for event searches due to the `rows` argument being set as part of the REST API call. the `_count()` and `_search()` functions both use the same arguments and the same urlobject when making the REST calls, except for `_search()` also specifying rows as 100 (the default batch size set via `args['rows'] = self._batch_size`).

To test I updated the _count() function to the following:

```python
    def _count(self):
        args = {"search_params": self._get_query_parameters()}

        log.debug("args: {}".format(str(args)))

        js = self._cb.post_object(
            self._doc_class.urlobject.format(self._cb.credentials.org_key), body=args
        ).json()

        self._total_results = int(js.get("response_header", {}).get("num_available", 0))
        self._count_valid = True

        print(f'_count args: {args}')
        print(f'_count json: {js.get("response_header")}\n')

        return self._total_results
```

And I updated the while loop in the `_search()` function to the following:

```python
       while still_querying:

            url = self._doc_class.urlobject.format(self._cb.credentials.org_key)
            resp = self._cb.post_object(url, body=args)
            result = resp.json()

            print(f'_search args: {args}')
            print(f'_search json: {result.get("response_header")}\n')

            self._total_results = result.get("response_header", {}).get("num_available", 0)
            self._count_valid = True

            results = result.get('docs', [])
```

Both print the arguments sent to the urlobject URL and the response header that is sent back from the REST API. Now we run the following script (please note this uses https://github.com/carbonblack/cbapi-python/pull/237 to fix the 403 issue with the `len()` function):

```python
from cbapi.psc.threathunter import CbThreatHunterAPI
from cbapi.psc.threathunter.models import Event

cb = CbThreatHunterAPI(profile='redlab2')
guid = 'ORGKEY-01319b03-00000cc4-00000000-1d64c634925f39c'
query_result = cb.select(Event).where(process_guid=guid).and_(event_type="crossproc")
print(f'len() count: {len(query_result)}')

i=0

for t in query_result:
    i+=1

print(f'Iteration count: {i}')
```

Output:
```
❯ python test.py
_count args: {'search_params': {'q': 'process_guid:ORGKEY\\-01319b03\\-00000cc4\\-00000000\\-1d64c634925f39c AND event_type:crossproc', 'cb.process_guid': 'ORGKEY-01319b03-00000cc4-00000000-1d64c634925f39c', 'fl': '*,parent_hash,parent_name,process_cmdline,backend_timestamp,device_external_ip,device_group,device_internal_ip,device_os,process_effective_reputation,process_reputation,ttp'}}
_count json: {'num_found': 13910, 'num_available': 1051, 'total_segments': 2145, 'processed_segments': 2145}

len() count: 1051
_search args: {'search_params': {'q': 'process_guid:ORGKEY\\-01319b03\\-00000cc4\\-00000000\\-1d64c634925f39c AND event_type:crossproc', 'cb.process_guid': 'ORGKEY-01319b03-00000cc4-00000000-1d64c634925f39c', 'fl': '*,parent_hash,parent_name,process_cmdline,backend_timestamp,device_external_ip,device_group,device_internal_ip,device_os,process_effective_reputation,process_reputation,ttp', 'rows': 100}}
_search json: {'num_found': 13910, 'num_available': 251, 'total_segments': 2145, 'processed_segments': 2145}
_search result count: 251
Iteration count: 251
```

The only difference between the `_count()` and the `_search()` function arguments is the `rows` value. It's set to 100 by default due to `args['rows'] = self._batch_size` and causes the platform to return a lower value for `num_available` than when `rows` is not specified at all.

In addition, even with the incorrect value for `num_available` (251 in the example above) the expectation would be that the while loop would loop at least 3 times as the requested row size is 100. 251 results would be loops of 100, 100, and 51. Instead however, the API returns all 251 results in a single go, resulting in a single iteration of the while loop. Iteration count near the end of the output shows how many actual results were returned. 251 in this case.

When we comment out `args['rows'] = self._batch_size`, in order to not send the `rows` argument in the `_search()` REST call, we see the following output instead:

```
❯ python test.py
_count args: {'search_params': {'q': 'process_guid:ORGKEY\\-01319b03\\-00000cc4\\-00000000\\-1d64c634925f39c AND event_type:crossproc', 'cb.process_guid': 'ORGKEY-01319b03-00000cc4-00000000-1d64c634925f39c', 'fl': '*,parent_hash,parent_name,process_cmdline,backend_timestamp,device_external_ip,device_group,device_internal_ip,device_os,process_effective_reputation,process_reputation,ttp'}}
_count json: {'num_found': 13910, 'num_available': 1051, 'total_segments': 2145, 'processed_segments': 2145}

len() count: 1051
_search args: {'search_params': {'q': 'process_guid:ORGKEY\\-01319b03\\-00000cc4\\-00000000\\-1d64c634925f39c AND event_type:crossproc', 'cb.process_guid': 'ORGKEY-01319b03-00000cc4-00000000-1d64c634925f39c', 'fl': '*,parent_hash,parent_name,process_cmdline,backend_timestamp,device_external_ip,device_group,device_internal_ip,device_os,process_effective_reputation,process_reputation,ttp'}}
_search json: {'num_found': 13910, 'num_available': 1051, 'total_segments': 2145, 'processed_segments': 2145}
_search result count: 500
_search args: {'search_params': {'q': 'process_guid:ORGKEY\\-01319b03\\-00000cc4\\-00000000\\-1d64c634925f39c AND event_type:crossproc', 'cb.process_guid': 'ORGKEY-01319b03-00000cc4-00000000-1d64c634925f39c', 'fl': '*,parent_hash,parent_name,process_cmdline,backend_timestamp,device_external_ip,device_group,device_internal_ip,device_os,process_effective_reputation,process_reputation,ttp'}, 'start': 501}
_search json: {'num_found': 13910, 'num_available': 1051, 'total_segments': 2145, 'processed_segments': 2145}
_search result count: 500
_search args: {'search_params': {'q': 'process_guid:ORGKEY\\-01319b03\\-00000cc4\\-00000000\\-1d64c634925f39c AND event_type:crossproc', 'cb.process_guid': 'ORGKEY-01319b03-00000cc4-00000000-1d64c634925f39c', 'fl': '*,parent_hash,parent_name,process_cmdline,backend_timestamp,device_external_ip,device_group,device_internal_ip,device_os,process_effective_reputation,process_reputation,ttp'}, 'start': 1001}
_search json: {'num_found': 13910, 'num_available': 1051, 'total_segments': 2145, 'processed_segments': 2145}
_search result count: 500
Iteration count: 1500
```

By not specifying a `rows` argument in the `_search()` request the platform now returns the same number of results as it does for the `_count()` function, and as expected it loops more frequently with a default result set of 500 per iteration. Surprisingly though, instead of returning 500, 500, 51 as indicated by `num_avialable`. It actually returns 3 batches of 500 results, effectively going beyond the reported number of `num_available`. The difference between `num_found` and `num_available` appears to be arbitrary and it's actually possible to retrieve more results than what `num_available` says are possible to be sent back from the platform.

In summary, `_search()` for events is broken in its current form and it appears to be in part at the side of the REST API in the way it handles requests that contain a `rows` parameter when submitting a search request.

1. Adding `rows` to a request should not impact `num_available` it should only change the number of results which are returned for that particular request.
2. When requesting results where the default result set size (i.e. 500 from our testing) goes beyond the value of `num_available` (e.g. starting at 1000 with a default result size of 500 when `num_available` is 1051) should yield 51 results and not 500. `num_available` should properly indicate the hard limit of retrievable results and not cause confusion by returning results beyond that value.

With the change to the v2 search at scale API, it's unclear how this will be resolved, but at this time event search results retrieved using the Python API are highly inconsistent and cause a lot of confusion. Especially because of the difference between what the `_count()` function returnes (and thus `.len()` on a query object returns) versus the actual results which are being retrieved (which is significantly less based on our testing: 251 vs 1051).

**Update**: for giggles I forced `args['rows'] = 100000000` in `_search()` and, as by magic, `num_available` changes to 13956 (which is actually all available results) and I'm able to retrieve all events in batches of 500. I've cut off the below output for brevity, but I hope this helps clarify the point that the value of `rows` causes issues in the API (and begs the question why `num_available` is set to such incredibly low values by default).

```
❯ python test.py
_count args: {'search_params': {'q': 'process_guid:N5DFVQXD\\-01319b03\\-00000cc4\\-00000000\\-1d64c634925f39c AND event_type:crossproc', 'cb.process_guid': 'N5DFVQXD-01319b03-00000cc4-00000000-1d64c634925f39c', 'fl': '*,parent_hash,parent_name,process_cmdline,backend_timestamp,device_external_ip,device_group,device_internal_ip,device_os,process_effective_reputation,process_reputation,ttp'}}
_count json: {'num_found': 13956, 'num_available': 1097, 'total_segments': 2145, 'processed_segments': 2145}

len() count: 1097
_search args: {'search_params': {'q': 'process_guid:ORGKEY\\-01319b03\\-00000cc4\\-00000000\\-1d64c634925f39c AND event_type:crossproc', 'cb.process_guid': 'ORGKEY-01319b03-00000cc4-00000000-1d64c634925f39c', 'fl': '*,parent_hash,parent_name,process_cmdline,backend_timestamp,device_external_ip,device_group,device_internal_ip,device_os,process_effective_reputation,process_reputation,ttp', 'rows': 100000000}}
_search json: {'num_found': 13956, 'num_available': 13956, 'total_segments': 2145, 'processed_segments': 2145}
_search result count: 500
_search args: {'search_params': {'q': 'process_guid:ORGKEY\\-01319b03\\-00000cc4\\-00000000\\-1d64c634925f39c AND event_type:crossproc', 'cb.process_guid': 'ORGKEY-01319b03-00000cc4-00000000-1d64c634925f39c', 'fl': '*,parent_hash,parent_name,process_cmdline,backend_timestamp,device_external_ip,device_group,device_internal_ip,device_os,process_effective_reputation,process_reputation,ttp', 'rows': 100000000}, 'start': 501}
_search json: {'num_found': 13956, 'num_available': 13956, 'total_segments': 2145, 'processed_segments': 2145}
_search result count: 500
_search args: {'search_params': {'q': 'process_guid:ORGKEY\\-01319b03\\-00000cc4\\-00000000\\-1d64c634925f39c AND event_type:crossproc', 'cb.process_guid': 'ORGKEY-01319b03-00000cc4-00000000-1d64c634925f39c', 'fl': '*,parent_hash,parent_name,process_cmdline,backend_timestamp,device_external_ip,device_group,device_internal_ip,device_os,process_effective_reputation,process_reputation,ttp', 'rows': 100000000}, 'start': 1001}
_search json: {'num_found': 13956, 'num_available': 13956, 'total_segments': 2145, 'processed_segments': 2145}
_search result count: 500
_search args: {'search_params': {'q': 'process_guid:ORGKEY\\-01319b03\\-00000cc4\\-00000000\\-1d64c634925f39c AND event_type:crossproc', 'cb.process_guid': 'ORGKEY-01319b03-00000cc4-00000000-1d64c634925f39c', 'fl': '*,parent_hash,parent_name,process_cmdline,backend_timestamp,device_external_ip,device_group,device_internal_ip,device_os,process_effective_reputation,process_reputation,ttp', 'rows': 100000000}, 'start': 1501}
_search json: {'num_found': 13956, 'num_available': 13956, 'total_segments': 2145, 'processed_segments': 2145}
_search result count: 500
_search args: {'search_params': {'q': 'process_guid:ORGKEY\\-01319b03\\-00000cc4\\-00000000\\-1d64c634925f39c AND event_type:crossproc', 'cb.process_guid': 'ORGKEY-01319b03-00000cc4-00000000-1d64c634925f39c', 'fl': '*,parent_hash,parent_name,process_cmdline,backend_timestamp,device_external_ip,device_group,device_internal_ip,device_os,process_effective_reputation,process_reputation,ttp', 'rows': 100000000}, 'start': 2001}
_search json: {'num_found': 13956, 'num_available': 13956, 'total_segments': 2145, 'processed_segments': 2145}
_search result count: 500
_search args: {'search_params': {'q': 'process_guid:ORGKEY\\-01319b03\\-00000cc4\\-00000000\\-1d64c634925f39c AND event_type:crossproc', 'cb.process_guid': 'ORGKEY-01319b03-00000cc4-00000000-1d64c634925f39c', 'fl': '*,parent_hash,parent_name,process_cmdline,backend_timestamp,device_external_ip,device_group,device_internal_ip,device_os,process_effective_reputation,process_reputation,ttp', 'rows': 100000000}, 'start': 2501}
_search json: {'num_found': 13956, 'num_available': 13956, 'total_segments': 2145, 'processed_segments': 2145}
_search result count: 500
_search args: {'search_params': {'q': 'process_guid:ORGKEY\\-01319b03\\-00000cc4\\-00000000\\-1d64c634925f39c AND event_type:crossproc', 'cb.process_guid': 'ORGKEY-01319b03-00000cc4-00000000-1d64c634925f39c', 'fl': '*,parent_hash,parent_name,process_cmdline,backend_timestamp,device_external_ip,device_group,device_internal_ip,device_os,process_effective_reputation,process_reputation,ttp', 'rows': 100000000}, 'start': 3001}
_search json: {'num_found': 13956, 'num_available': 13956, 'total_segments': 2145, 'processed_segments': 2145}
_search result count: 500
...
```

**Comment by avanbrunt-cb** [comment]
@crahan We've identified the cause of the change for process events. The request is synchronous in nature but takes time for the searchers to come back so the initial call triggers the fetch of the data. Subsequent calls will return more data until it finishes fetching the results at which point the full set is fetch-able until the num_available amount.

A recent PR https://github.com/carbonblack/cbapi-python/pull/244 has exposed the segment response body headers that indicate the progress of the searchers returning results. This can be used to make the same query until total and processed segments are equivalent. 

```
query = cb.select(Event).where("process_guid:xxxx")
query[0]
while query._total_segments != query._processed_segments:
    query[0]

for item in query:
    print(item)
```

We plan on addressing this better with CBAPI in a future release.

**Comment by crahan** [comment]
Thank you. I'll test this asap. In the meantime I've updated https://github.com/carbonblack/cbapi-python/pull/241 to reflect the proper response fields for the v2 search API.

**Comment by crahan** [comment]
I ran a second test to see how many unique event GUIDs I was able to pull when setting `rows` to 100000000 and here's the outcome:

```python
from cbapi.psc.threathunter import CbThreatHunterAPI
from cbapi.psc.threathunter.models import Event

cb = CbThreatHunterAPI(profile='redlab2')
guid = 'ORGKEY-01319b03-00000cc4-00000000-1d64c634925f39c'
query_result = cb.select(Event).where(process_guid=guid).and_(event_type="crossproc")

print(f'len() count: {len(query_result)}')

# workaround
query_result[0]
while query_result._total_segments != query_result._processed_segments:
    print("Hitting the workaround!")
    query_result[0]

i = 0

results = []

for t in query_result:
    results.append(t.event_guid)
    i += 1

print(f'Iteration count: {i}')
print("Original", len(results))
print("Unique", len(set(results)))
```

And the output for this is (I removed the `_search()` print statements):

```
❯ python test4.py
_count args: {'search_params': {'q': 'process_guid:ORGKEY\\-01319b03\\-00000cc4\\-00000000\\-1d64c634925f39c AND event_type:crossproc', 'cb.process_guid': 'ORGKEY-01319b03-00000cc4-00000000-1d64c634925f39c', 'fl': '*,parent_hash,parent_name,process_cmdline,backend_timestamp,device_external_ip,device_group,device_internal_ip,device_os,process_effective_reputation,process_reputation,ttp'}}
_count json: {'num_found': 29955, 'num_available': 1601, 'total_segments': 4638, 'processed_segments': 4638}

len() count: 1601
Iteration count: 28000
Original 28000
Unique 507
```

So with `num_available` reported as being 1601,  `total_segments` and `processed_segments` being the same from the start (i.e. 4638), I was able to iterate 28000 times (so the API returned at least 28000 results). But there were only 507 unique event GUIDs. With 1601 `num_available` results, how come only 507 are unique?

**Comment by avanbrunt-cb** [comment]
For context, total_segments and processed_segments will only differ when the data is not cached. When the searchers have to fetch the data it takes time for them to return so if the same process_guid request is being made and the data is already cached they will remain equivalent. The workaround will only apply for when a new process_guid that is not cached is requested. The num_available and num_found properties will grow as more segments are processed which is why a count prior to the workaround will return a smaller number.

Additionally, the process search v1 events route doesn't support rows and start that is a process search v2 implementation instead the query params need to be `cb.row_count` and `cb.start_row`. This is a bug with the current implementation which is actually handled by your v2 PR.

As for the change in num_available based on the rows, I was unable to reproduce the same behavior however your second test would behave as expected given that the following break would not happen until 28000.
```
if current >= self._total_results:
                break
```
Does the same issue exist on the your process search v2 PR?

**Comment by crahan** [comment]
For sure! v2 does appear to work a lot more reliably based on initial testing of event searches. As noted on https://github.com/carbonblack/cbapi-python/pull/241:

"We are also able to retrieve more or even all events for a particular Process object (versus only a handful of events with the v1 event search API). Retrieval of event data now correctly uses batches of 100 events per iteration and it will loop until the value of num_available has been reached (fixing the issues reported in #239)."

Getting the changes for API v2 process searches into master would help us significantly and resolve this issue. I'm sure that Cb would prefer to move to v2 as well and not spend time debugging and fixing issues related to the legacy v1 search API.

Just so I understand though. When you mention "this is a bug with the current implementation". Is the below `_search()` snippet from the current `master` branch what you are referring to? That `rows` is being submitted as part of the v1 event search request instead of `cb.row_count`, which causes things to potentially break?

```    
def _search(self, start=0, rows=0):
        # iterate over total result set, 100 at a time
        args = self._get_query_parameters()
        self._validate(args)

        if start != 0:
            args['start'] = start
        args['rows'] = self._batch_size

        args = {"search_params": args}
```

**Comment by avanbrunt-cb** [comment]
We are reviewing the active PRs and will hopefully get some merged soon. 

I ran some API requests with the rows option but was unable to reproduce the same behavior. Looking at how the events call is made from our UI console it utilizes cb.row_count instead. Ideally, an unknown property would be ignored and the rows should not affect num_available only the amount of rows that are returned from num_available. Since this affects the older v1 route, we will refer to the new route for a fix.

**Comment by crahan** [comment]
Sounds good, thanks for confirming. Can we leave this issue open until #241 is reviewed and merged (or an alternative v2 search implementation has been added to cbapi)?

**Comment by avanbrunt-cb** [comment]
#241 Is merged and will be included in the 1.7.0 release
